units/src/amount.rs

```diff
@@ -368,33 +505,44 @@ impl std::error::Error for PossiblyConfusingDenominationError {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
 }
 
-fn is_too_precise(s: &str, precision: usize) -> bool {
+/// Returns `Some(position)` if the precision is not supported.
+///
+/// The position indicates the first digit that is too precise.
+fn is_too_precise(s: &str, precision: usize) -> Option<usize> {
     match s.find('.') {
-        Some(pos) =>
-            s[(pos + 1)..].chars().any(|d| d != '0')
-                || precision >= pos
-                || s[..pos].chars().rev().take(precision).any(|d| d != '0'),
-        None => precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0'),
+        Some(pos) if precision >= pos => { Some(0) },
+        Some(pos) => {
+            s[..pos].char_indices().rev().take(precision).find(|(_, d)| *d != '0').map(|(i, _)| i)
+                .or_else(|| {
+                    s[(pos + 1)..].char_indices().find(|(_, d)| *d != '0').map(|(i, _)| i + pos + 1)
+                })
+        },
+        None if precision >= s.len() => { Some(0) },
+        None => {
+            s.char_indices().rev().take(precision).find(|(_, d)| *d != '0').map(|(i, _)| i)
+        },
     }
 }
 
+const INPUT_STRING_LEN_LIMIT: usize = 50;
```

Comment:
Can we have a code comment about where this limit comes from please, for me to learn and for the next guy :)

Reply:
I personally don't know but I believe it's some kind of DoS protection. All reasonable amounts fit 50 characters. Sure you could have a bunch of zeros in front and after dot but it's silly. Though honestly I'd prefer the limit to not exist from consumer standpoint.